modules/phase_field/src/userobjects/EBSDReader.C

```diff
@@ -331,56 +331,94 @@ EBSDReader::buildNodeWeightMaps()
-      return new EBSDPointDataGrain();
+      ret_val = new EBSDPointDataGrain();
+      break;
     case 4: // phase
-      return new EBSDPointDataPhase();
+      ret_val = new EBSDPointDataPhase();
+      break;
     case 5: // symmetry
-      return new EBSDPointDataSymmetry();
+      ret_val = new EBSDPointDataSymmetry();
+      break;
     case 6: // op
-      return new EBSDPointDataOp();
+      ret_val = new EBSDPointDataOp();
+      break;
+    default:
+    {
+      // check for custom columns
+      for (unsigned int i = 0; i < _custom_columns; ++i)
+        if (field_name == "CUSTOM" + Moose::stringify(i))
+        {
+          ret_val = new EBSDPointDataCustom(i);
+          break;
+        }
+    }
   }
 
-  // check for custom columns
-  for (unsigned int i = 0; i < _custom_columns; ++i)
-    if (field_name == "CUSTOM" + Moose::stringify(i))
-      return new EBSDPointDataCustom(i);
+  // If ret_val was not set by any of the above cases, throw an error.
+  if (!ret_val)
+    mooseError("Error:  Please input supported EBSD_param");
 
-  mooseError("Error:  Please input supported EBSD_param");
+  // If we made it here, wrap up the the ret_val in a
+  // MooseSharedPointer and ship it off.
+  return MooseSharedPointer<EBSDAccessFunctors::EBSDPointDataFunctor>(ret_val);
 }
 
-EBSDAccessFunctors::EBSDAvgDataFunctor *
+MooseSharedPointer<EBSDAccessFunctors::EBSDAvgDataFunctor>
 EBSDReader::getAvgDataAccessFunctor(const MooseEnum & field_name) const
 {
+  EBSDAccessFunctors::EBSDAvgDataFunctor * ret_val = NULL;
+
   switch (field_name)
   {
     case 0: // phi1
-      return new EBSDAvgDataPhi1();
+      ret_val = new EBSDAvgDataPhi1();
+      break;
     case 1: // phi
-      return new EBSDAvgDataPhi();
+      ret_val = new EBSDAvgDataPhi();
+      break;
     case 2: // phi2
-      return new EBSDAvgDataPhi2();
+      ret_val = new EBSDAvgDataPhi2();
+      break;
     case 3: // phase
-      return new EBSDAvgDataPhase();
+      ret_val = new EBSDAvgDataPhase();
+      break;
     case 4: // symmetry
-      return new EBSDAvgDataSymmetry();
+      ret_val = new EBSDAvgDataSymmetry();
+      break;
+    default:
+    {
+      // check for custom columns
+      for (unsigned int i = 0; i < _custom_columns; ++i)
+        if (field_name == "CUSTOM" + Moose::stringify(i))
+        {
+          ret_val = new EBSDAvgDataCustom(i);
+          break;
+        }
+    }
   }
 
-  // check for custom columns
-  for (unsigned int i = 0; i < _custom_columns; ++i)
-    if (field_name == "CUSTOM" + Moose::stringify(i))
-      return new EBSDAvgDataCustom(i);
+  // If ret_val was not set by any of the above cases, throw an error.
+  if (!ret_val)
+    mooseError("Error:  Please input supported EBSD_param");
 
-  mooseError("Error:  Please input supported EBSD_param");
+  // If we made it here, wrap up the the ret_val in a
+  // MooseSharedPointer and ship it off.
+  return MooseSharedPointer<EBSDAccessFunctors::EBSDAvgDataFunctor>(ret_val);
```

Comment:
We do.  The first thing I tried was just returning "new Foo" as in the old code, but I got a bunch of "could not convert pointer-to-derived-type" to shared_ptr<base-type> error messages.  It would have been nice if that would have just worked...